### PR TITLE
syncer(dm): use DML library

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tiflow/issues/3242

### What is changed and how it works?

DM side switched to the DML library of https://github.com/pingcap/tiflow/pull/4376

### Check List 

Tests 

 - pass original test

Code changes

Side effects

Related changes

### Release note 

```release-note
`None`.
```

### CODE REVIEW

**Comment by lance6716** [comment]
GBK integration test has passed succesful, I'll split this PR into smaller ones to make it easier to review

**Comment by lance6716** [comment]
a good CI https://ci2.pingcap.net/blue/organizations/jenkins/dm_ghpr_integration_test/detail/dm_ghpr_integration_test/2843/pipeline

**Comment by lance6716** [comment]
many changes are moving unit tests 😂

**Comment by lance6716** [comment]
unit tests and integration tests passed. Most changes are to cope with existing unit tests and delete repeat codes.

/cc @GMHDBJD @glorv @Ehco1996

File: dm/syncer/causality_test.go [hunk 1]
```diff
@@ -100,124 +83,44 @@ func (s *testSyncerSuite) TestCasuality(c *C) {
...
-	}), IsTrue)
+	}, 3*time.Second, 100*time.Millisecond)
 
 	for _, op := range results {
 		job := <-causalityCh
-		c.Assert(job.tp, Equals, op)
-	}
-}
-
-func (s *testSyncerSuite) TestCasualityWithPrefixIndex(c *C) {
```
Comment (GMHDBJD) [inline dm/syncer/causality_test.go:149]: Does this test have been cover?

File: dm/syncer/compactor.go [hunk 2]
```diff
@@ -192,34 +190,30 @@ func (c *compactor) compactJob(j *job) {
...
 		}
-	case insert:
-		if prevJob.tp == del {
+	case sqlmodel.RowChangeInsert:
+		if prevJob.dml.Type() == sqlmodel.RowChangeDelete {
 			// DELETE + INSERT => INSERT with safemode
-			j.dml.safeMode = true
+			j.safeMode = true
+			skipReduce = true
 		}
```
Comment (GMHDBJD) [inline dm/syncer/compactor.go:206]: Seems Delete aslo can skip reduce?

File: dm/syncer/compactor.go [hunk 3]
```diff
@@ -192,34 +190,30 @@ func (c *compactor) compactJob(j *job) {
...
-	case del:
-		// do nothing because anything + DELETE => DELETE
+	}
```
Comment (GMHDBJD) [inline dm/syncer/compactor.go:207]: Why not also skip reduce RowChangeDelete?

File: dm/syncer/compactor.go [hunk 4]
```diff
@@ -89,23 +89,21 @@ func (c *compactor) run() {
...
+			if !j.dml.HasNotNullUniqueIdx() {
 				c.buffer = append(c.buffer, j)
 				continue
 			}
 
 			// if update job update its identify keys, turn it into delete + insert
-			if j.dml.op == update && j.dml.updateIdentify() {
-				delDML, insertDML := updateToDelAndInsert(j.dml)
+			if j.dml.IsIdentityUpdated() {
+				delDML, insertDML := j.dml.Split()
```
Comment (GMHDBJD) [inline dm/syncer/compactor.go:102]: How about SplitUpdate()

**Comment by GMHDBJD** [comment]
rest LGTM

**Review by GMHDBJD — COMMENTED** [review]
rest LGTM

File: dm/syncer/compactor.go (hunk 3 above)
Comment (lance6716) [reply dm/syncer/compactor.go:207]: we only run compactor for row changes whose PR/UK doesn't change, I think it's both OK. 

And I want to apply Reduce to all combinations of row change so I want to reduce the presence of `skipReduce`. If we implement multi-row version of UPDATE in future, I think we can also remove the `// DELETE + INSERT => INSERT with safemode` extra path

**Review by GMHDBJD — APPROVED** [review]
LGTM

**Comment by lance6716** [comment]
ptal @Ehco1996 @glorv @lichunzhu 

CI has passed in previous commits

File: dm/syncer/job.go [hunk 5]
```diff
@@ -228,6 +233,24 @@ func newConflictJob(workerCount int) *job {
...
+// newCompactJob is only used for metrics.
+func newCompactJob(targetTable *filter.Table) *job {
+	return &job{
+		tp:          compact,
+		targetTable: targetTable,
+	}
+}
+
+// newDummyJob is only used in tests.
+func newDummyJob(tp opType, targetTable *filter.Table, ddls ...string) *job {
```
Comment (glorv) [inline dm/syncer/job.go:245]: Nit: We can move this function to _test.go since it is a package private function.

**Review by glorv — APPROVED** [review]
LGTM

File: dm/syncer/dml.go [hunk 6]
```diff
@@ -15,67 +15,32 @@ package syncer
...
-		return "delete"
-	case insertOnDuplicateDML:
-		return "insert on duplicate update"
-	case replaceDML:
-		return "replace"
-	}
-	return
-}
-
 // genDMLParam stores pruned columns, data as well as the original columns, data, index.
```
Comment (lichunzhu) [inline dm/syncer/dml.go:36]: We should update this comment

**Pull request merged**